src/Propulsion.SqlStreamStore/StreamReader.fs

```diff
@@ -0,0 +1,226 @@
+                )
+
+            { stream = StreamName.internalParseSafe msg.StreamId; event = event }
+
+    type Stats(logger: ILogger, ?statsInterval: TimeSpan) =
+
+        let statsInterval = defaultArg statsInterval (TimeSpan.FromSeconds(30.))
+
+        let mutable batchFirstPosition = 0L
+        let mutable batchLastPosition = 0L
+        let mutable batchCaughtUp = false
+
+        let mutable pagesRead = 0
+        let mutable pagesEmpty = 0
+        let mutable recentPagesRead = 0
+        let mutable recentPagesEmpty = 0
+
+        let mutable currentBatches = 0
+        let mutable maxBatches = 0
+
+        let mutable lastCommittedPosition = 0L
+
+        let report () =
+            logger.Information(
+                "Pages Read {pagesRead} Empty {pagesEmpty} | Recent Read {recentPagesRead} Empty {recentPagesEmpty} | Position Read {batchLastPosition} Committed {lastCommittedPosition} | Caught up {caughtUp} | cur {cur} / max {max}",
+                pagesRead, pagesEmpty, recentPagesRead, recentPagesEmpty, batchLastPosition, lastCommittedPosition, batchCaughtUp, currentBatches, maxBatches)
+
+        member this.UpdateBatch (batch: InternalBatch) =
+            batchFirstPosition <- batch.firstPosition
+            batchLastPosition <- batch.lastPosition
+            batchCaughtUp <- batch.isEnd
+
+            pagesRead <- pagesRead + 1
+            recentPagesRead <- recentPagesRead + 1
+
+        member this.UpdateEmptyPage () =
+            pagesEmpty <- pagesEmpty + 1
+            recentPagesEmpty <- recentPagesEmpty + 1
+
+        member this.UpdateCommitedPosition(pos) =
+            lastCommittedPosition <- pos
+
+        member this.UpdateCurMax(cur, max) =
+            currentBatches <- cur
+            maxBatches <- max
+
+        member this.Start =
+            async {
+                let! ct = Async.CancellationToken
+                while not ct.IsCancellationRequested do
+                    report ()
+                    recentPagesRead <- 0
+                    recentPagesEmpty <- 0
+                    do! Async.Sleep statsInterval
+            }
+
+    [<RequireQualifiedAccess>]
+    [<NoComparison>]
+    type Work =
+        | TakeInitial
+        | Page of ReadAllPage
+        | TakeNext of ReadAllPage
+
+    type SubmitBatchHandler =
+        // ingester submit method: epoch * checkpoint * items -> write result
+        int64 * Async<unit> * seq<Propulsion.Streams.StreamEvent<byte[]>> -> Async<int*int>
+
+type StreamReader
+    (
+        logger : ILogger,
+        store: IStreamStore,
+        ledger: ICheckpointer,
+        submitBatch: SubmitBatchHandler,
+        streamId,
+        consumerGroup,
+        ?maxBatchSize: int,
+        ?sleepInterval: TimeSpan,
+        ?statsInterval: TimeSpan,
+        ?stopAfterInterval: TimeSpan
+    ) =
+
+    let maxBatchSize = defaultArg maxBatchSize 100
+    let sleepInterval = defaultArg sleepInterval (TimeSpan.FromSeconds(5.))
```

Comment:
5s is a really long default - maybe make it mandatory so caller makes it explicit?

Reply:
Yeah, it's long. Propose another default?

Reply:
I'd propose to remove defaulting and let the value live in Program.fs in a template
For Es, the TailInterval default I happen to use is -1s
(Also that name makes it clearer what the sleep is for and when it happens)

Reply:
Chances are I derived the 1s from the default value for the CFP one: https://github.com/jet/propulsion/search?q=feedpolldelay&unscoped_q=feedpolldelay
(that, and the sqlstreamstore one, could both do with a comment as to what the default value is)